File: clients/c/test-mumbojumbo-client.c
#include "mumbojumbo-client.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <sodium.h>

#define ASSERT(cond, msg) do { \
    if (!(cond)) { \
        fprintf(stderr, "FAILED: %s\n", msg); \
        return 0; \
    } \
} while(0)

#define RUN_TEST(test) do { \
    printf("Running %s... ", #test); \
    if (test()) { \
        printf("PASSED\n"); \
        passed++; \
    } else { \
        printf("FAILED\n"); \
        failed++; \
    } \
    total++; \
} while(0)

// Test: Key Parsing
static int test_parse_valid_hex_key(void) {
    uint8_t key[32];
    char valid_key[72];
    strcpy(valid_key, "mj_pub_");
    for (int i = 0; i < 64; i++) {
        strcat(valid_key, "a");
    }

    ASSERT(parse_key_hex(valid_key, key) == 0, "parse_key_hex should succeed");
    ASSERT(key[0] == 0xaa, "first byte should be 0xaa");
    return 1;
}

static int test_reject_key_without_prefix(void) {
    uint8_t key[32];
    char invalid_key[65];
    for (int i = 0; i < 64; i++) {
        invalid_key[i] = 'a';
    }
    invalid_key[64] = '\0';

    ASSERT(parse_key_hex(invalid_key, key) != 0, "should reject key without prefix");
    return 1;
}

static int test_reject_key_with_wrong_length(void) {
    uint8_t key[32];
    ASSERT(parse_key_hex("mj_pub_aabbccdd", key) != 0, "should reject short key");
    return 1;
}

// Test: Base32 Encoding
static int test_base32_encode_basic(void) {
    const uint8_t data[] = "hello";
    char *encoded = base32_encode(data, 5);
    ASSERT(encoded != NULL, "encoding should succeed");
    ASSERT(strcmp(encoded, "nbswy3dp") == 0, "encoding should match expected");
    free(encoded);
    return 1;
}

static int test_base32_encode_empty(void) {
    char *encoded = base32_encode(NULL, 0);
    ASSERT(encoded != NULL, "encoding empty should succeed");
    ASSERT(strlen(encoded) == 0, "encoding empty should return empty string");
    free(encoded);
    return 1;
}

static int test_base32_encode_lowercase(void) {
    const uint8_t data[] = "TEST";
    char *encoded = base32_encode(data, 4);
    ASSERT(encoded != NULL, "encoding should succeed");

    for (size_t i = 0; encoded[i]; i++) {
        ASSERT(encoded[i] == tolower(encoded[i]), "should be lowercase");
    }

    free(encoded);
    return 1;
}

static int test_base32_encode_no_padding(void) {
    const uint8_t data[] = "hello world";
    char *encoded = base32_encode(data, 11);
    ASSERT(encoded != NULL, "encoding should succeed");
    ASSERT(strchr(encoded, '=') == NULL, "should not contain padding");
    free(encoded);
    return 1;
}

// Test: DNS Label Splitting
static int test_split_to_labels_short(void) {
    size_t count;
    char **labels = split_to_labels("abc", DNS_LABEL_MAX_LEN, &count);
    ASSERT(count == 1, "should have 1 label");
    ASSERT(strcmp(labels[0], "abc") == 0, "label should match");
    free_labels(labels, count);
    return 1;
}

static int test_split_to_labels_exact_63(void) {
    char data[64];
    memset(data, 'a', 63);
    data[63] = '\0';

    size_t count;
    char **labels = split_to_labels(data, DNS_LABEL_MAX_LEN, &count);
    ASSERT(count == 1, "should have 1 label");
    ASSERT(strlen(labels[0]) == 63, "label should be 63 chars");
    free_labels(labels, count);
    return 1;
}

static int test_split_to_labels_long(void) {
    char data[101];
    memset(data, 'a', 100);
    data[100] = '\0';

    size_t count;
    char **labels = split_to_labels(data, DNS_LABEL_MAX_LEN, &count);
    ASSERT(count == 2, "should have 2 labels");
    ASSERT(strlen(labels[0]) == 63, "first label should be 63 chars");
    ASSERT(strlen(labels[1]) == 37, "second label should be 37 chars");
    free_labels(labels, count);
    return 1;
}

static int test_split_to_labels_empty(void) {
    size_t count;
    char **labels = split_to_labels("", DNS_LABEL_MAX_LEN, &count);
    ASSERT(count == 0, "should have 0 labels");
    ASSERT(labels == NULL, "should return NULL");
    return 1;
}

// Test: Fragment Creation
static int test_create_basic_fragment(void) {
    const uint8_t frag_data[] = "test";
    uint8_t *frag;
    size_t frag_len;

    ASSERT(create_fragment(100, 0, 1, frag_data, 4, &frag, &frag_len) == 0, "fragment creation should succeed");
    ASSERT(frag_len == HEADER_SIZE + 4, "fragment length should be correct");

    uint16_t packet_id = (frag[0] << 8) | frag[1];
    uint32_t frag_index = (frag[2] << 24) | (frag[3] << 16) | (frag[4] << 8) | frag[5];
    uint32_t frag_count = (frag[6] << 24) | (frag[7] << 16) | (frag[8] << 8) | frag[9];
    uint16_t data_len = (frag[10] << 8) | frag[11];

    ASSERT(packet_id == 100, "packet_id should be 100");
    ASSERT(frag_index == 0, "frag_index should be 0");
    ASSERT(frag_count == 1, "frag_count should be 1");
    ASSERT(data_len == 4, "data_len should be 4");
    ASSERT(memcmp(frag + HEADER_SIZE, frag_data, 4) == 0, "fragment data should match");

    free(frag);
    return 1;
}

static int test_create_empty_fragment(void) {
    uint8_t *frag;
    size_t frag_len;

    ASSERT(create_fragment(1, 0, 1, NULL, 0, &frag, &frag_len) == 0, "empty fragment creation should succeed");
    ASSERT(frag_len == HEADER_SIZE, "empty fragment length should be HEADER_SIZE");

    uint16_t data_len = (frag[10] << 8) | frag[11];
    ASSERT(data_len == 0, "data_len should be 0");

    free(frag);
    return 1;
}

static int test_reject_oversized_fragment(void) {
    uint8_t oversized[MAX_FRAG_DATA_LEN + 1];
    uint8_t *frag;
    size_t frag_len;

    ASSERT(create_fragment(1, 0, 1, oversized, sizeof(oversized), &frag, &frag_len) != 0,
           "should reject oversized fragment");
    return 1;
}

static int test_reject_invalid_frag_index(void) {
    const uint8_t data[] = "test";
    uint8_t *frag;
    size_t frag_len;

    ASSERT(create_fragment(1, 5, 5, data, 4, &frag, &frag_len) != 0,
           "should reject invalid frag_index");
    return 1;
}

// Test: Data Fragmentation
static int test_fragment_small_data(void) {
    const uint8_t data[] = "small";
    uint8_t **fragments;
    size_t *frag_lens;
    size_t count;

    ASSERT(fragment_data(data, 5, MAX_FRAG_DATA_LEN, &fragments, &frag_lens, &count) == 0,
           "fragmentation should succeed");
    ASSERT(count == 1, "should have 1 fragment");
    ASSERT(frag_lens[0] == 5, "fragment should be 5 bytes");
    ASSERT(memcmp(fragments[0], data, 5) == 0, "fragment data should match");

    free_fragments(fragments, frag_lens, count);
    return 1;
}

static int test_fragment_exact_max_size(void) {
    uint8_t data[MAX_FRAG_DATA_LEN];
    memset(data, 0x42, MAX_FRAG_DATA_LEN);

    uint8_t **fragments;
    size_t *frag_lens;
    size_t count;

    ASSERT(fragment_data(data, MAX_FRAG_DATA_LEN, MAX_FRAG_DATA_LEN, &fragments, &frag_lens, &count) == 0,
           "fragmentation should succeed");
    ASSERT(count == 1, "should have 1 fragment");
    ASSERT(frag_lens[0] == MAX_FRAG_DATA_LEN, "fragment should be MAX_FRAG_DATA_LEN");

    free_fragments(fragments, frag_lens, count);
    return 1;
}

static int test_fragment_overflow(void) {
    uint8_t data[MAX_FRAG_DATA_LEN + 10];
    memset(data, 0x42, sizeof(data));

    uint8_t **fragments;
    size_t *frag_lens;
    size_t count;

    ASSERT(fragment_data(data, sizeof(data), MAX_FRAG_DATA_LEN, &fragments, &frag_lens, &count) == 0,
           "fragmentation should succeed");
    ASSERT(count == 2, "should have 2 fragments");
    ASSERT(frag_lens[0] == MAX_FRAG_DATA_LEN, "first fragment should be MAX_FRAG_DATA_LEN");
    ASSERT(frag_lens[1] == 10, "second fragment should be 10");

    free_fragments(fragments, frag_lens, count);
    return 1;
}

static int test_fragment_empty_data(void) {
    uint8_t **fragments;
    size_t *frag_lens;
    size_t count;

    ASSERT(fragment_data(NULL, 0, MAX_FRAG_DATA_LEN, &fragments, &frag_lens, &count) == 0,
           "empty fragmentation should succeed");
    ASSERT(count == 1, "should have 1 fragment");
    ASSERT(frag_lens[0] == 0, "fragment should be empty");

    free_fragments(fragments, frag_lens, count);
    return 1;
}

// Test: Client API
static int test_initialize_client(void) {
    uint8_t pubkey[32];
    memset(pubkey, 0xaa, 32);

    MumbojumboClient *client = mumbojumbo_client_new(pubkey, ".asd.qwe", MAX_FRAG_DATA_LEN);
    ASSERT(client != NULL, "client creation should succeed");
    ASSERT(strcmp(client->domain, ".asd.qwe") == 0, "domain should match");
    ASSERT(client->max_fragment_size == MAX_FRAG_DATA_LEN, "max_fragment_size should match");

    mumbojumbo_client_free(client);
    return 1;
}

static int test_auto_prepend_dot_to_domain(void) {
    uint8_t pubkey[32];
    memset(pubkey, 0xaa, 32);

    MumbojumboClient *client = mumbojumbo_client_new(pubkey, "asd.qwe", MAX_FRAG_DATA_LEN);
    ASSERT(client != NULL, "client creation should succeed");
    ASSERT(strcmp(client->domain, ".asd.qwe") == 0, "domain should have prepended dot");

    mumbojumbo_client_free(client);
    return 1;
}

static int test_generate_queries_without_sending(void) {
    uint8_t pubkey[32];
    memset(pubkey, 0xaa, 32);

    MumbojumboClient *client = mumbojumbo_client_new(pubkey, ".asd.qwe", MAX_FRAG_DATA_LEN);
    ASSERT(client != NULL, "client creation should succeed");

    char **queries;
    size_t count;

    ASSERT(mumbojumbo_generate_queries(client, (const uint8_t *)"test", 4, &queries, &count) == 0,
           "query generation should succeed");
    ASSERT(count == 1, "should have 1 query");
    ASSERT(strstr(queries[0], ".asd.qwe") != NULL, "query should contain domain");

    free_queries(queries, count);
    mumbojumbo_client_free(client);
    return 1;
}

int main(void) {
    if (sodium_init() < 0) {
        fprintf(stderr, "Failed to initialize libsodium\n");
        return 1;
    }

    int total = 0, passed = 0, failed = 0;

    printf("\n=== Mumbojumbo C Client Tests ===\n\n");

    // Key Parsing Tests
    printf("Key Parsing Tests:\n");
    RUN_TEST(test_parse_valid_hex_key);
    RUN_TEST(test_reject_key_without_prefix);
    RUN_TEST(test_reject_key_with_wrong_length);

    // Base32 Encoding Tests
    printf("\nBase32 Encoding Tests:\n");
    RUN_TEST(test_base32_encode_basic);
    RUN_TEST(test_base32_encode_empty);
    RUN_TEST(test_base32_encode_lowercase);
    RUN_TEST(test_base32_encode_no_padding);

    // DNS Label Splitting Tests
    printf("\nDNS Label Splitting Tests:\n");
    RUN_TEST(test_split_to_labels_short);
    RUN_TEST(test_split_to_labels_exact_63);
    RUN_TEST(test_split_to_labels_long);
    RUN_TEST(test_split_to_labels_empty);

    // Fragment Creation Tests
    printf("\nFragment Creation Tests:\n");
    RUN_TEST(test_create_basic_fragment);
    RUN_TEST(test_create_empty_fragment);
    RUN_TEST(test_reject_oversized_fragment);
    RUN_TEST(test_reject_invalid_frag_index);

    // Data Fragmentation Tests
    printf("\nData Fragmentation Tests:\n");
    RUN_TEST(test_fragment_small_data);
    RUN_TEST(test_fragment_exact_max_size);
    RUN_TEST(test_fragment_overflow);
    RUN_TEST(test_fragment_empty_data);

    // Client API Tests
    printf("\nClient API Tests:\n");
    RUN_TEST(test_initialize_client);
    RUN_TEST(test_auto_prepend_dot_to_domain);
    RUN_TEST(test_generate_queries_without_sending);

    printf("\n=== Test Results ===\n");
    printf("Total:  %d\n", total);
    printf("Passed: %d\n", passed);
    printf("Failed: %d\n\n", failed);

    return failed > 0 ? 1 : 0;
}

File: clients/nodejs/test-mumbojumbo-client.js
#!/usr/bin/env node
/**
 * Comprehensive tests for Node.js mumbojumbo client
 *
 * Mirrors the 43 tests from the Python implementation.
 * Run with: node --test test-mumbojumbo-client.js
 */

const { test, describe } = require('node:test');
const assert = require('node:assert/strict');
const nacl = require('tweetnacl');
const sealedBox = require('tweetnacl-sealedbox-js');
const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

const {
  base32Encode,
  splitToLabels,
  createFragment,
  encryptFragment,
  createDnsQuery,
  fragmentData,
  MumbojumboClient,
  MAX_FRAG_DATA_LEN,
  DNS_LABEL_MAX_LEN
} = require('./mumbojumbo-client.js');

// Test fixtures
const TEST_KEYPAIR = nacl.box.keyPair();
const TEST_SERVER_PUBKEY = TEST_KEYPAIR.publicKey;
const TEST_SERVER_PRIVKEY = TEST_KEYPAIR.secretKey;
const TEST_DOMAIN = '.asd.qwe';

/**
 * Decrypt using NaCl SealedBox
 */
function decryptSealedBox(ciphertext, publicKey, privateKey) {
  return sealedBox.open(ciphertext, publicKey, privateKey);
}

/**
 * Base32 decode (lowercase, no padding)
 */
function base32Decode(encoded) {
  const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
  let bits = 0;
  let value = 0;
  const output = [];

  for (const char of encoded.toUpperCase()) {
    const index = alphabet.indexOf(char);
    if (index === -1) continue;

    value = (value << 5) | index;
    bits += 5;

    if (bits >= 8) {
      output.push((value >>> (bits - 8)) & 0xFF);
      bits -= 8;
    }
  }

  return Buffer.from(output);
}

/**
 * Run CLI command and capture output
 */
async function runCli(args, stdinData = null) {
  return new Promise((resolve, reject) => {
    const child = spawn('node', ['./mumbojumbo-client.js', ...args]);

    const stdout = [];
    const stderr = [];

    child.stdout.on('data', (data) => stdout.push(data));
    child.stderr.on('data', (data) => stderr.push(data));

    child.on('close', (code) => {
      resolve({
        code,
        stdout: Buffer.concat(stdout).toString(),
        stderr: Buffer.concat(stderr).toString()
      });
    });

    child.on('error', reject);

    if (stdinData) {
      child.stdin.write(stdinData);
      child.stdin.end();
    }
  });
}

// ============================================================================
// Test: Key Parsing (via MumbojumboClient constructor)
// ============================================================================

describe('Key Parsing', () => {
  test('parse valid hex key via constructor', () => {
    const validKey = 'mj_cli_' + 'a'.repeat(64);
    const client = new MumbojumboClient(validKey, TEST_DOMAIN);

    assert.ok(client.serverPubkey instanceof Uint8Array);
    assert.equal(client.serverPubkey.length, 32);
    assert.equal(client.serverPubkey[0], 0xaa);
  });

  test('reject key without mj_cli_ prefix', () => {
    assert.throws(() => {
      new MumbojumboClient('a'.repeat(64), TEST_DOMAIN);
    }, /must start with "mj_cli_"/);
  });

  test('reject key with wrong hex length', () => {
    assert.throws(() => {
      new MumbojumboClient('mj_cli_' + 'a'.repeat(32), TEST_DOMAIN);
    }, /Invalid hex key length/);
  });

  test('reject key with invalid hex characters', () => {
    assert.throws(() => {
      new MumbojumboClient('mj_cli_' + 'z'.repeat(64), TEST_DOMAIN);
    }, /Invalid hex characters/);
  });
});

// ============================================================================
// Test: Base32 Encoding
// ============================================================================

describe('Base32 Encoding', () => {
  test('encode basic data', () => {
    const data = Buffer.from('hello');
    const encoded = base32Encode(data);

    assert.equal(encoded, 'nbswy3dp');
  });

  test('encode empty data', () => {
    const encoded = base32Encode(Buffer.alloc(0));
    assert.equal(encoded, '');
  });

  test('encode produces lowercase', () => {
    const data = Buffer.from('TEST');
    const encoded = base32Encode(data);

    assert.equal(encoded, encoded.toLowerCase());
  });

  test('encode has no padding', () => {
    const data = Buffer.from('hello world');
    const encoded = base32Encode(data);

    assert.ok(!encoded.includes('='));
  });

  test('encode/decode round-trip', () => {
    const original = Buffer.from('Round trip test data!');
    const encoded = base32Encode(original);
    const decoded = base32Decode(encoded);

    assert.deepEqual(decoded, original);
  });
});

// ============================================================================
// Test: DNS Label Splitting
// ============================================================================

describe('DNS Label Splitting', () => {
  test('split short string', () => {
    const labels = splitToLabels('abc');
    assert.deepEqual(labels, ['abc']);
  });

  test('split exactly 63 chars', () => {
    const data = 'a'.repeat(63);
    const labels = splitToLabels(data);
    assert.deepEqual(labels, [data]);
  });

  test('split long string', () => {
    const data = 'a'.repeat(100);
    const labels = splitToLabels(data);

    assert.equal(labels.length, 2);
    assert.equal(labels[0].length, 63);
    assert.equal(labels[1].length, 37);
  });

  test('split empty string', () => {
    const labels = splitToLabels('');
    assert.deepEqual(labels, []);
  });

  test('split with custom max length', () => {
    const labels = splitToLabels('abcdefgh', 3);
    assert.deepEqual(labels, ['abc', 'def', 'gh']);
  });
});

// ============================================================================
// Test: Fragment Creation
// ============================================================================

describe('Fragment Creation', () => {
  test('create basic fragment', () => {
    const fragData = Buffer.from('test');
    const frag = createFragment(100, 0, 1, fragData);

    assert.equal(frag.length, 18 + 4);

    // Parse header (18 bytes: u64 packet_id, u32 frag_index, u32 frag_count, u16 frag_data_len)
    const packetIdHigh = frag.readUInt32BE(0);
    const packetIdLow = frag.readUInt32BE(4);
    const packetId = packetIdHigh * 0x100000000 + packetIdLow;
    const fragIndex = frag.readUInt32BE(8);
    const fragCount = frag.readUInt32BE(12);
    const dataLen = frag.readUInt16BE(16);

    assert.equal(packetId, 100);
    assert.equal(fragIndex, 0);
    assert.equal(fragCount, 1);
    assert.equal(dataLen, 4);
    assert.deepEqual(frag.slice(18), fragData);
  });

  test('create multi-fragment packet', () => {
    const fragData = Buffer.from('part1');
    const frag = createFragment(200, 2, 5, fragData);

    const fragIndex = frag.readUInt32BE(8);
    const fragCount = frag.readUInt32BE(12);

    assert.equal(fragIndex, 2);
    assert.equal(fragCount, 5);
  });

  test('create empty fragment', () => {
    const frag = createFragment(1, 0, 1, Buffer.alloc(0));

    assert.equal(frag.length, 18);
    assert.equal(frag.readUInt16BE(16), 0);
  });

  test('reject oversized fragment', () => {
    const oversized = Buffer.alloc(MAX_FRAG_DATA_LEN + 1);

    assert.throws(() => {
      createFragment(1, 0, 1, oversized);
    }, /Fragment data too large/);
  });

  test('reject invalid packet_id', () => {
    assert.throws(() => {
      createFragment(-1, 0, 1, Buffer.from('test'));
    }, /packet_id out of range/);

    // u64 max is Number.MAX_SAFE_INTEGER in JavaScript
    assert.throws(() => {
      createFragment(Number.MAX_SAFE_INTEGER + 1, 0, 1, Buffer.from('test'));
    }, /packet_id out of range/);
  });

  test('reject invalid frag_index', () => {
    assert.throws(() => {
      createFragment(1, -1, 1, Buffer.from('test'));
    }, /Invalid frag_index/);

    assert.throws(() => {
      createFragment(1, 5, 5, Buffer.from('test'));
    }, /Invalid frag_index/);
  });

  test('support u32 frag_count', () => {
    const largeCount = 0xFFFFFFFF;
    const frag = createFragment(1, 0, largeCount, Buffer.from('test'));

    const fragCount = frag.readUInt32BE(12);
    assert.equal(fragCount, largeCount);
  });

  test('support u32 frag_index', () => {
    const largeIndex = 1000000;
    const frag = createFragment(1, largeIndex, largeIndex + 1, Buffer.from('test'));

    const fragIndex = frag.readUInt32BE(8);
    assert.equal(fragIndex, largeIndex);
  });
});

// ============================================================================
// Test: Data Fragmentation
// ============================================================================

describe('Data Fragmentation', () => {
  test('fragment small data', () => {
    const data = Buffer.from('small');
    const fragments = fragmentData(data);

    assert.equal(fragments.length, 1);
    assert.deepEqual(fragments[0], data);
  });

  test('fragment exactly max size', () => {
    const data = Buffer.alloc(MAX_FRAG_DATA_LEN);
    const fragments = fragmentData(data);

    assert.equal(fragments.length, 1);
    assert.equal(fragments[0].length, MAX_FRAG_DATA_LEN);
  });

  test('fragment overflow', () => {
    const data = Buffer.alloc(MAX_FRAG_DATA_LEN + 10);
    const fragments = fragmentData(data);

    assert.equal(fragments.length, 2);
    assert.equal(fragments[0].length, MAX_FRAG_DATA_LEN);
    assert.equal(fragments[1].length, 10);
  });

  test('fragment empty data', () => {
    const fragments = fragmentData(Buffer.alloc(0));

    assert.equal(fragments.length, 1);
    assert.equal(fragments[0].length, 0);
  });

  test('fragment large data', () => {
    const data = Buffer.alloc(250);
    const fragments = fragmentData(data);

    const expectedCount = Math.ceil(250 / MAX_FRAG_DATA_LEN);
    assert.equal(fragments.length, expectedCount);
  });
});

// ============================================================================
// Test: Encryption
// ============================================================================

describe('Encryption', () => {
  test('encrypt fragment', () => {
    const plaintext = Buffer.from('secret data');
    const encrypted = encryptFragment(plaintext, TEST_SERVER_PUBKEY);

    assert.ok(encrypted.length > plaintext.length);
    assert.equal(encrypted.length, plaintext.length + 48);
  });

  test('encrypt/decrypt round-trip', () => {
    const plaintext = Buffer.from('test message');

    const encrypted = encryptFragment(plaintext, TEST_SERVER_PUBKEY);
    const decrypted = decryptSealedBox(encrypted, TEST_SERVER_PUBKEY, TEST_SERVER_PRIVKEY);

    assert.deepEqual(Buffer.from(decrypted), plaintext);
  });

  test('encrypt empty data', () => {
    const encrypted = encryptFragment(Buffer.alloc(0), TEST_SERVER_PUBKEY);

    assert.equal(encrypted.length, 48);
  });

  test('encryption produces different outputs', () => {
    const plaintext = Buffer.from('same data');

    const enc1 = encryptFragment(plaintext, TEST_SERVER_PUBKEY);
    const enc2 = encryptFragment(plaintext, TEST_SERVER_PUBKEY);

    assert.notDeepEqual(enc1, enc2);
  });
});

// ============================================================================
// Test: DNS Query Creation
// ============================================================================

describe('DNS Query Creation', () => {
  test('create basic DNS query', () => {
    const encrypted = Buffer.from('a'.repeat(10));
    const query = createDnsQuery(encrypted, TEST_DOMAIN);

    assert.ok(query.endsWith(TEST_DOMAIN));
  });

  test('create query splits long base32', () => {
    const encrypted = Buffer.alloc(100);
    const query = createDnsQuery(encrypted, TEST_DOMAIN);

    const labels = query.replace(TEST_DOMAIN, '').split('.');
    for (const label of labels) {
      assert.ok(label.length <= DNS_LABEL_MAX_LEN);
    }
  });

  test('create query from empty data', () => {
    const query = createDnsQuery(Buffer.alloc(0), TEST_DOMAIN);
    assert.equal(query, TEST_DOMAIN);
  });
});

// ============================================================================
// Test: MumbojumboClient Class
// ============================================================================

describe('MumbojumboClient Class', () => {
  test('initialize client', () => {
    const client = new MumbojumboClient(TEST_SERVER_PUBKEY, TEST_DOMAIN);

    assert.ok(client.serverPubkey);
    assert.equal(client.domain, TEST_DOMAIN);
    assert.equal(client.maxFragmentSize, MAX_FRAG_DATA_LEN);
  });

  test('initialize with Buffer public key', () => {
    const bufferKey = Buffer.from(TEST_SERVER_PUBKEY);
    const client = new MumbojumboClient(bufferKey, TEST_DOMAIN);

    assert.ok(client.serverPubkey instanceof Uint8Array);
  });

  test('auto-prepend dot to domain', () => {
    const client = new MumbojumboClient(TEST_SERVER_PUBKEY, 'asd.qwe');
    assert.equal(client.domain, '.asd.qwe');
  });

  test('generate queries without sending', async () => {
    const client = new MumbojumboClient(TEST_SERVER_PUBKEY, TEST_DOMAIN);
    const queries = await client.generateQueries(Buffer.from('test'));

    assert.ok(Array.isArray(queries));
    assert.equal(queries.length, 1);
    assert.ok(queries[0].endsWith(TEST_DOMAIN));
  });

  test('generate queries only', async () => {
    const client = new MumbojumboClient(TEST_SERVER_PUBKEY, TEST_DOMAIN);
    const queries = await client.generateQueries(Buffer.from('test'));

    assert.ok(Array.isArray(queries));
    assert.equal(queries.length, 1);
    assert.ok(queries[0].endsWith(TEST_DOMAIN));
  });

  test('packet ID increments', async () => {
    const client = new MumbojumboClient(TEST_SERVER_PUBKEY, TEST_DOMAIN);

    const queries1 = await client.generateQueries(Buffer.from('msg1'));
    const queries2 = await client.generateQueries(Buffer.from('msg2'));

    assert.notEqual(queries1[0], queries2[0]);
  });

  test('packet ID wraps at MAX_SAFE_INTEGER', () => {
    const client = new MumbojumboClient(TEST_SERVER_PUBKEY, TEST_DOMAIN);
    client._nextPacketId = Number.MAX_SAFE_INTEGER - 1;

    const id1 = client._getNextPacketId();
    const id2 = client._getNextPacketId();

    assert.equal(id1, Number.MAX_SAFE_INTEGER - 1);
    assert.equal(id2, 0);
  });

  test('multi-fragment message', async () => {
    const client = new MumbojumboClient(TEST_SERVER_PUBKEY, TEST_DOMAIN);
    const largeData = Buffer.alloc(MAX_FRAG_DATA_LEN * 3);

    const queries = await client.generateQueries(largeData);
    assert.equal(queries.length, 3);
  });
});

// ============================================================================
// Test: End-to-End Flow
// ============================================================================

describe('End-to-End Flow', () => {
  test('encrypt and decrypt single fragment', async () => {
    const client = new MumbojumboClient(TEST_SERVER_PUBKEY, TEST_DOMAIN);
    const message = Buffer.from('Hello World');

    const queries = await client.generateQueries(message);
    assert.equal(queries.length, 1);

    const query = queries[0];
    const base32Part = query.replace(TEST_DOMAIN, '').replace(/\./g, '');
    const encrypted = base32Decode(base32Part);
    const decrypted = decryptSealedBox(encrypted, TEST_SERVER_PUBKEY, TEST_SERVER_PRIVKEY);

    const header = Buffer.from(decrypted.slice(0, 18));
    const data = Buffer.from(decrypted.slice(18));

    const dataLen = header.readUInt16BE(16);
    assert.equal(dataLen, message.length);
    assert.deepEqual(data, message);
  });

  test('encrypt and decrypt multi-fragment', async () => {
    const client = new MumbojumboClient(TEST_SERVER_PUBKEY, TEST_DOMAIN);
    const message = Buffer.alloc(MAX_FRAG_DATA_LEN * 2 + 10);
    message.fill(0x42);

    const queries = await client.generateQueries(message);
    assert.equal(queries.length, 3);

    const fragments = [];

    for (const query of queries) {
      const base32Part = query.replace(TEST_DOMAIN, '').replace(/\./g, '');
      const encrypted = base32Decode(base32Part);
      const decrypted = decryptSealedBox(encrypted, TEST_SERVER_PUBKEY, TEST_SERVER_PRIVKEY);

      const header = Buffer.from(decrypted.slice(0, 18));
      const data = Buffer.from(decrypted.slice(18));

      const packetIdHigh = header.readUInt32BE(0);
      const packetIdLow = header.readUInt32BE(4);
      const packetId = packetIdHigh * 0x100000000 + packetIdLow;
      const fragIndex = header.readUInt32BE(8);
      const fragCount = header.readUInt32BE(12);
      const dataLen = header.readUInt16BE(16);

      fragments.push({ packetId, fragIndex, fragCount, data: data.subarray(0, dataLen) });
    }

    assert.equal(fragments.length, 3);
    assert.equal(fragments[0].fragCount, 3);
    assert.equal(fragments[0].fragIndex, 0);
    assert.equal(fragments[1].fragIndex, 1);
    assert.equal(fragments[2].fragIndex, 2);

    const reassembled = Buffer.concat(fragments.map(f => f.data));
    assert.deepEqual(reassembled, message);
  });
});

// ============================================================================
// Test: CLI Interface
// ============================================================================

describe('CLI Interface', () => {
  const testKeyHex = 'mj_cli_' + Buffer.from(TEST_SERVER_PUBKEY).toString('hex');

  test('show help', async () => {
    const result = await runCli(['--help']);

    assert.equal(result.code, 0);
    assert.ok(result.stdout.includes('Usage:'));
    assert.ok(result.stdout.includes('--key'));
    assert.ok(result.stdout.includes('--domain'));
  });

  test('missing required arguments', async () => {
    const result = await runCli([]);

    assert.notEqual(result.code, 0);
    assert.ok(result.stderr.includes('required'));
  });

  test('read from stdin', async () => {
    const result = await runCli([
      '-k', testKeyHex,
      '-d', TEST_DOMAIN
    ], 'test message');

    assert.equal(result.code, 0);
    assert.ok(result.stdout.includes(TEST_DOMAIN));
  });

  test('read from file', async () => {
    const tmpFile = path.join(__dirname, 'test-input.tmp');
    fs.writeFileSync(tmpFile, 'file content');

    try {
      const result = await runCli([
        '-k', testKeyHex,
        '-d', TEST_DOMAIN,
        '-f', tmpFile
      ]);

      assert.equal(result.code, 0);
      assert.ok(result.stdout.includes(TEST_DOMAIN));
    } finally {
      fs.unlinkSync(tmpFile);
    }
  });

  test('verbose mode', async () => {
    const result = await runCli([
      '-k', testKeyHex,
      '-d', TEST_DOMAIN,
      '-v'
    ], 'test');

    assert.equal(result.code, 0);
    assert.ok(result.stderr.includes('bytes'));
  });

  test('invalid key format', async () => {
    const result = await runCli([
      '-k', 'invalid_key',
      '-d', TEST_DOMAIN
    ], 'test');

    assert.notEqual(result.code, 0);
    assert.ok(result.stderr.includes('Error'));
  });
});

console.log('\n✅ All tests defined. Run with: node --test test-mumbojumbo-client.js\n');
